File: Python/day03/AOC.py
import re
def part1(input):
    return sum([int(match[0]) * int(match[1]) for match in re.findall(r"mul\((\d+),(\d+)\)", input)])

# The input had new lines which threw off the regex, so I had to use re.DOTALL
# used re.sub to remove the don't() part of the string
def part2(input):
    return part1(re.sub(r"don't\(\).*?(?=do\(\)|$)", "", input, flags=re.DOTALL))
    

def main():
    input = open("input.txt", "r").read()
    part1_ans = part1(input)
    part2_ans = part2(input)
    print(f"Part 1: {part1_ans}")
    print(f"Part 2: {part2_ans}")

if __name__ == "__main__":
    main()

File: Go/day03/AOC.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func part1(input string) int {
	sum := 0
	for _, c := range regexp.MustCompile(`mul\((\d+),(\d+)\)`).FindAllStringSubmatch(input, -1) {
		a, b := c[1], c[2]
		num1, _ := strconv.Atoi(a)
		num2, _ := strconv.Atoi(b)
		sum = num1 * num2
	}
	return sum
}

func part2(input string) int {
	enabled := true
	sum := 0
	for _, c := range regexp.MustCompile(`don't|do|mul\(\d+,\d+\)`).FindAllStringSubmatch(input, -1) {
		if c[0] == "do" {
			enabled = true
		} else if c[0] == "don't" {
			enabled = false
		} else if enabled {
			sum += part1(c[0])
		}
	}
	return sum
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := string(input)
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
